app/routes.py

```diff
@@ -1 +1,298 @@
+    return task
+
+
+# wave 2 
+# get task by asc and desc order
+@task_bp.route("", methods =["GET"])
+def sort_title_asc_and_desc():
+    sort_by = request.args.get("sort")
+
+    if sort_by == "asc":
+        tasks = Task.query.order_by(Task.title.asc()).all()
+    elif sort_by == "desc":
+        tasks = Task.query.order_by(Task.title.desc()).all()
+    else:
+        tasks = Task.query.all() # this line could save one function for get_all_tasks
+    
+
+    response =[task.to_dict() for task in tasks]
+
+    return jsonify(response), 200
+
+
+# wave 3 
+# mark complete
+@task_bp.route("/<task_id>/mark_complete", methods = ["PATCH"])
+def mark_complete(task_id):
+
+    task = Task.query.get(task_id)
+    if not task:
+        return jsonify({"msg": "task not found"}), 404
+    
+    if not task.completed_at:
+        task.completed_at = datetime.utcnow()
+        task.is_complete = True 
+
+        db.session.commit()
+
+        url = "https://slack.com/api/chat.postMessage"
+        token = os.environ.get("SLACKBOT_TOKEN")
+        data ={"channel": "task-notifications",
+               "text": f"Someone just completed a task {task.title}",
+               "token": token
+               }
+        response = requests.post(url, data = data)
+
+
+        # this is the part with slack module:
+        # message = f"Someone just completed a task {task.title}"
+        # client = WebClient(token=os.environ.get("SLACKBOT_TOKEN"))
+        # response = client.chat_postMessage(channel="task-notifications", text=message)
+
+         
+    return jsonify({"task":task.to_dict()}), 200
+
+
+# mark_imcomplete
+@task_bp.route("/<task_id>/mark_incomplete", methods = ["PATCH"])
+def mark_imcomplete(task_id):
+    task = Task.query.get(task_id)
+    if not task:
+        return jsonify({"msg": "task not found"}), 404
+    
+    task.completed_at = None
+    task.is_complete = False
+
+    db.session.commit()
+
+    return jsonify({"task":task.to_dict()}), 200
+
+
+
+
+# GOAL
+goal_bp = Blueprint("goal", __name__,url_prefix = "/goals")
+
+# wave 5
+# create a goal
+@goal_bp.route("", methods = ["POST"])
+def create_goal():
+    
+    response_body = request.get_json()
+    if "title" not in response_body:
+        return jsonify({"details": "Invalid data"}), 400
+    new_goal = Goal.from_goal_dict(response_body)
+    
+    db.session.add(new_goal)
+    db.session.commit()
+  
+    return jsonify({
+                    "goal":{
+                    "id":new_goal.goal_id,
+                    "title":new_goal.title
+    }}), 201
+
+
+# helper function
+def validate_goal(model,goal_id):
+    try:
+        goal_id = int(goal_id)
+    except ValueError:
+        return abort(make_response({"msg": "id is invalide input"}, 400))
+    
+    # return Task.query.get_or_404(task_id_num)
+ 
+    goal = model.query.get(goal_id)
+    if goal is None:
+        abort(make_response({"msg": "Goal not found"}, 404))
+    return goal
```

Comment:
This helper function does almost the exact same thing as the `validate_task` helper function. There is a way to make it so that you only need one function to do both of these things.